pennylane_lightning/core/src/simulators/lightning_kokkos/gates/BasicGateFunctors.hpp

```diff
@@ -64,105 +115,249 @@ template <class PrecisionT, class FuncT> class applyNC1Functor {
+              const std::size_t num_qubits,
+              const std::vector<std::size_t> &controlled_wires,
+              const std::vector<bool> &controlled_values,
+              const std::vector<std::size_t> &wires, const bool inverse = false,
+              [[maybe_unused]] const std::vector<PrecisionT> &params = {}) {
+    const Kokkos::complex<PrecisionT> shift =
+        (inverse) ? Kokkos::complex<PrecisionT>{0.0, -1.0}
+                  : Kokkos::complex<PrecisionT>{0.0, 1.0};
+    auto core_function =
         KOKKOS_LAMBDA(Kokkos::View<Kokkos::complex<PrecisionT> *> arr,
                       const std::size_t i0, const std::size_t i1) {
-            [[maybe_unused]] const auto i0_ = i0;
-            const Kokkos::complex<PrecisionT> v0 = arr(i0);
-            const Kokkos::complex<PrecisionT> v1 = arr(i1);
-            arr(i0) = M_SQRT1_2 * v0 +
-                      M_SQRT1_2 * v1; // NOLINT(readability-magic-numbers)
-            arr(i1) = M_SQRT1_2 * v0 +
-                      -M_SQRT1_2 * v1; // NOLINT(readability-magic-numbers)
-        });
+        [[maybe_unused]] const auto i0_ = i0;
+        arr(i1) *= shift;
+    };
+    if (controlled_wires.empty()) {
+        applyNC1Functor<PrecisionT, decltype(core_function), false>(
+            ExecutionSpace{}, arr_, num_qubits, wires, core_function);
+    } else {
+        applyNC1Functor<PrecisionT, decltype(core_function), true>(
+            ExecutionSpace{}, arr_, num_qubits, controlled_wires,
+            controlled_values, wires, core_function);
+    }
 }
 
 template <class ExecutionSpace, class PrecisionT>
 void applyS(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
             const std::size_t num_qubits, const std::vector<std::size_t> &wires,
             const bool inverse = false,
             [[maybe_unused]] const std::vector<PrecisionT> &params = {}) {
+    applyNCS<ExecutionSpace, PrecisionT>(arr_, num_qubits, {}, {}, wires,
+                                         inverse);
+}
+
+template <class ExecutionSpace, class PrecisionT>
+void applyNCT(Kokkos::View<Kokkos::complex<PrecisionT> *> arr_,
+              const std::size_t num_qubits,
+              const std::vector<std::size_t> &controlled_wires,
+              const std::vector<bool> &controlled_values,
+              const std::vector<std::size_t> &wires, const bool inverse = false,
+              [[maybe_unused]] const std::vector<PrecisionT> &params = {}) {
     const Kokkos::complex<PrecisionT> shift =
-        (inverse) ? Kokkos::complex<PrecisionT>{0.0, -1.0}
-                  : Kokkos::complex<PrecisionT>{0.0, 1.0};
-    applyNC1Functor(
-        ExecutionSpace{}, arr_, num_qubits, wires,
+        (inverse) ? Kokkos::conj(exp(Kokkos::complex<PrecisionT>(
+                        0, static_cast<PrecisionT>(M_PI_4))))
+                  : exp(Kokkos::complex<PrecisionT>(
+                        0, static_cast<PrecisionT>(M_PI_4)));
```

Comment:
My only concern is to use the correct math function because is not the same using `std::sin` as `sin` and Kokkos takes care of this. Also, `sin`, `cos`, `exp`... only works with `double` then if we use `float` instead of `double` will be a slowdown because the compiler needs to convert the `float` -> `double` -> math func -> `double` - `float`. However, `std` the math function is templated to `sinf` or `sin` or `sinl` ([ref](https://en.cppreference.com/w/cpp/numeric/math/sin))

Reply:
I will use `Kokkos::` for `kokkos::complex` types and in parallel regions, and `std::` for otherwise